packages/frontend/src/scripts/favicon-dot.ts

```diff
@@ -0,0 +1,132 @@
+	private ctx: CanvasRenderingContext2D | null = null;
+	private faviconImage: HTMLImageElement | null = null;
+	private faviconEl: HTMLLinkElement | undefined;
+	private hasLoaded: Promise<void> | undefined;
+
+	constructor() {
+		this.canvas = document.createElement('canvas');
+	}
+
+	/**
+	 * Must be called before calling any other functions
+	 */
+	public async setup() {
+		const element: HTMLLinkElement = await this.getOrMakeFaviconElement();
+
+		this.faviconEl = element;
+		this.ctx = this.canvas.getContext('2d');
+
+		this.faviconImage = document.createElement('img');
+
+		this.hasLoaded = new Promise((resolve, reject) => {
+			if (!this.faviconImage) {
+				reject('Failed to create favicon img element');
+				return;
+			}
+
+			this.faviconImage.addEventListener('load', () => {
+				this.canvas.width = this.faviconImage!.width;
+				this.canvas.height = this.faviconImage!.height;
+				resolve();
+			});
+			this.faviconImage.addEventListener('error', () => {
+				reject('Failed to create favicon img element');
+			});
+		});
+
+		this.faviconImage.src = this.faviconEl.href;
+	}
+
+	private async getOrMakeFaviconElement(): Promise<HTMLLinkElement> {
+		return new Promise((resolve, reject) => {
+			const favicon = document.querySelector<HTMLLinkElement>('link[rel=icon]') ?? this.createFaviconElem();
+			favicon.addEventListener('load', () => {
+				resolve(favicon);
+			});
+
+			favicon.onerror = () => {
+				reject('Failed to load favicon');
+			};
+			resolve(favicon);
+		});
+	}
+
+	private createFaviconElem() {
+		const newLink = document.createElement('link');
+		newLink.setAttribute('rel', 'icon');
+		newLink.setAttribute('href', '/favicon.ico');
+		newLink.setAttribute('type', 'image/x-icon');
+
+		document.head.appendChild(newLink);
+		return newLink;
+	}
+
+	private drawIcon() {
+		if (!this.ctx || !this.faviconImage) return;
+		this.ctx.clearRect(0, 0, this.canvas.width, this.canvas.height);
+		this.ctx.drawImage(this.faviconImage, 0, 0, this.faviconImage.width, this.faviconImage.height);
+	}
+
+	private drawDot() {
+		if (!this.ctx || !this.faviconImage) return;
+		const radius = Math.min(this.faviconImage.width, this.faviconImage.height) / 6;
+		const computedStyle = getComputedStyle(document.documentElement);
+
+		// 元の画像を丸く切り抜く
+		this.ctx.globalCompositeOperation = 'destination-out';
+		this.ctx.beginPath();
+		this.ctx.arc(this.canvas.width - radius, this.canvas.height - radius, radius * 1.5, 0, 2 * Math.PI);
+		this.ctx.closePath();
+		this.ctx.fill();
+
+		// 丸い部分を描画
+		this.ctx.globalCompositeOperation = 'source-over';
+		this.ctx.beginPath();
+		this.ctx.arc(this.canvas.width - radius, this.canvas.height - radius, radius, 0, 2 * Math.PI);
+		this.ctx.fillStyle = tinycolor(computedStyle.getPropertyValue('--error') || '#ec4137').toHexString();
+		this.ctx.closePath();
+		this.ctx.fill();
+	}
+
+	private setFavicon() {
+		if (this.faviconEl) this.faviconEl.href = this.canvas.toDataURL('image/png');
```

Comment:
you probably should mention in some documentation that this only works if the favicon URL is on the same origin as the frontend, or if the favicon origin explicitly allows CORS
https://activitypub.software/TransFem-org/Documentation/-/merge_requests/23/diffs#54a895f7109f42da43fc5248ff6ef8d51275d93b this is what we're adding to our docs